File: app/src/main/java/org/peercast/core/PeerCastFragment.java
package org.peercast.core;


import android.content.ActivityNotFoundException;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v7.app.ActionBar;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.view.ContextMenu;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ExpandableListView;
import android.widget.TextView;
import android.widget.Toast;

import static org.peercast.core.PeerCastServiceController.MSG_GET_APPLICATION_PROPERTIES;
import static org.peercast.core.PeerCastServiceController.MSG_GET_CHANNELS;
import static org.peercast.core.PeerCastServiceController.MSG_GET_STATS;

/**
 * (c) 2015, T Yoshizawa
 * Dual licensed under the MIT or GPL licenses.
 */
public class PeerCastFragment extends Fragment implements
        PeerCastServiceController.OnServiceResultListener {

    static final String TAG = "PeerCastFragment";
    private int mRunningPort;

    private GuiListAdapter mListAdapter;


    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        return inflater.inflate(R.layout.peercast_fragment, container, false);
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        setHasOptionsMenu(true);
        AppCompatActivity act = (AppCompatActivity) getActivity();
        ActionBar actionBar = act.getSupportActionBar();
        actionBar.setDisplayHomeAsUpEnabled(false);
        actionBar.setTitle(R.string.app_name);

        ExpandableListView lv = (ExpandableListView) getView().findViewById(R.id.vListChannel);
        mListAdapter = new GuiListAdapter(getContext());
        lv.setAdapter(mListAdapter);

        registerForContextMenu(lv);
    }


    public PeerCastActivity getActivity2() {
        return (PeerCastActivity) getActivity();
    }

    /**
     * @see PeerCastActivity#onServiceResult(int, Bundle)
     */
    @Override
    public void onServiceResult(int what, Bundle data) {
        if (getView() == null)
            return;

        switch (what) {
            case MSG_GET_CHANNELS:
                mListAdapter.setChannels(Channel.fromNativeResult(data));
                break;

            case MSG_GET_STATS:
                TextView vBandwidth = (TextView) getView().findViewById(R.id.vBandwidth);
                Stats stats = Stats.fromNativeResult(data);
                String s = String.format("R: %.1fkbps S: %.1fkbps   Port: %d ", stats.getInBytes() / 1000f * 8, stats.getOutBytes() / 1000f * 8, mRunningPort);
                vBandwidth.setText(s);
                break;

            case MSG_GET_APPLICATION_PROPERTIES:
                mRunningPort = data.getInt("port");
                break;
        }
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        inflater.inflate(R.menu.peercast_menu, menu);
    }


    @Override
    public void onPrepareOptionsMenu(Menu menu) {
        boolean started = mRunningPort > 0;
        //menu.findItem(R.id.menu_html_settings).setEnabled(started);
        menu.findItem(R.id.menu_settings).setEnabled(started);
    }

    @Override
    public void onCreateContextMenu(ContextMenu menu, View v, ContextMenu.ContextMenuInfo menuInfo) {
        ExpandableListView.ExpandableListContextMenuInfo info = (ExpandableListView.ExpandableListContextMenuInfo) menuInfo;

        int type = ExpandableListView.getPackedPositionType(info.packedPosition);
        int gPos = ExpandableListView.getPackedPositionGroup(info.packedPosition);
        int cPos = ExpandableListView.getPackedPositionChild(info.packedPosition);

        if (type == ExpandableListView.PACKED_POSITION_TYPE_GROUP) {
            // チャンネル
            getActivity().getMenuInflater()
                    .inflate(R.menu.channel_context_menu, menu);
            Channel ch = mListAdapter.getGroup(gPos);
            menu.setHeaderTitle(ch.getInfo()
                    .getName());
            MenuItem mKeep = menu.findItem(R.id.menu_ch_keep);
            mKeep.setChecked(ch.isStayConnected());
        } else if (type == ExpandableListView.PACKED_POSITION_TYPE_CHILD) {
            // サーヴァント
            getActivity().getMenuInflater()
                    .inflate(R.menu.servent_context_menu, menu);
            Servent svt = mListAdapter.getChild(gPos, cPos);
            menu.setHeaderTitle(svt.getHost());
        }
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        Intent settingIntent = new Intent(getContext(), SettingActivity.class);

        switch (item.getItemId()) {
            case R.id.menu_remove_all_channel:
                for (Channel ch : mListAdapter.getChannels()) {
                    if (ch.getLocalRelays() + ch.getLocalListeners() == 0)
                        getActivity2().getPeerCastServiceController().disconnectChannel(ch.getChannel_ID());
                }
                return true;

            case R.id.menu_html_settings:
                settingIntent.putExtra(SettingActivity.LAUNCH_HTML, true);
            case R.id.menu_settings: {
                settingIntent.putExtra(SettingActivity.PORT, mRunningPort);
                startActivityForResult(settingIntent, 0);
                //overridePendingTransition(0, 0);
                return true;
            }

            default:
                return super.onOptionsItemSelected(item);
        }
    }

    @Override
    public boolean onContextItemSelected(MenuItem item) {
        PeerCastServiceController controller = getActivity2().getPeerCastServiceController();

        ExpandableListView.ExpandableListContextMenuInfo info = (ExpandableListView.ExpandableListContextMenuInfo) item.getMenuInfo();
        int type = ExpandableListView.getPackedPositionType(info.packedPosition);
        int gPos = ExpandableListView.getPackedPositionGroup(info.packedPosition);
        int cPos = ExpandableListView.getPackedPositionChild(info.packedPosition);

        Channel ch = (Channel) mListAdapter.getGroup(gPos);

        switch (item.getItemId()) {

            case R.id.menu_ch_disconnect:
                Log.i(TAG, "Disconnect channel: " + ch);
                controller.disconnectChannel(ch.getChannel_ID());
                return true;

            case R.id.menu_ch_keep:
                Log.i(TAG, "Keep channel: " + ch);
                controller.setChannelKeep(ch.getChannel_ID(), !item.isChecked());
                return true;

            case R.id.menu_ch_play:
                Uri u = getStreamUri(ch);
                Intent intent = new Intent(Intent.ACTION_VIEW, u);
                try {
                    showToast(u.toString());
                    startActivity(intent);
                } catch (ActivityNotFoundException e) {
                    getActivity2().showAlertDialog(R.string.t_error, e.getLocalizedMessage(), false);
                }
                return true;

            case R.id.menu_ch_bump:
                Log.i(TAG, "Bump channel: " + ch);
                controller.bumpChannel(ch.getChannel_ID());
                return true;

            case R.id.menu_svt_disconnect:
                //直下切断
                Servent svt = (Servent) mListAdapter.getChild(gPos, cPos);
                Log.i(TAG, "Disconnect servent: " + svt);
                controller.disconnectServent(svt.getServent_ID());
                return true;

            default:
                return super.onContextItemSelected(item);
        }
    }

    private void showToast(String msg) {
        Toast.makeText(getContext(), msg, Toast.LENGTH_LONG)
                .show();
    }

    /**
     * チャンネル再生用のURL
     */
    public Uri getStreamUri(Channel ch) {
        return Util.getStreamUrl(ch, mRunningPort);
    }


}

File: app/src/main/java/org/peercast/core/PeerCastFragment.kt
package org.peercast.core


import android.content.ActivityNotFoundException
import android.os.Bundle
import android.view.*
import android.widget.ExpandableListView
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import androidx.fragment.app.Fragment
import androidx.lifecycle.lifecycleScope
import com.google.android.material.snackbar.Snackbar
import kotlinx.coroutines.delay
import kotlinx.coroutines.isActive
import org.koin.android.ext.android.inject
import org.koin.androidx.viewmodel.ext.android.sharedViewModel
import org.peercast.core.databinding.PeercastFragmentBinding
import org.peercast.core.lib.LibPeerCast
import org.peercast.core.lib.PeerCastRpcClient
import org.peercast.core.preferences.AppPreferences
import timber.log.Timber

/**
 * @author (c) 2014-2019, T Yoshizawa
 * @licenses Dual licensed under the MIT or GPL licenses.
 */
class PeerCastFragment : Fragment() {
    private val viewModel by sharedViewModel<AppViewModel>()
    private val appPrefs by inject<AppPreferences>()
    private val listAdapter = GuiListAdapter()

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        viewLifecycleOwner.lifecycleScope.launchWhenStarted {
            while (isActive){
                listAdapter.channels = viewModel.getActiveChannels()
                delay(8_000)
            }
        }

        return PeercastFragmentBinding.inflate(inflater, container, false).also {
            it.viewModel = viewModel
            it.lifecycleOwner = viewLifecycleOwner
            it.vListChannel.setAdapter(listAdapter)
            registerForContextMenu(it.vListChannel)
        }.root
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        setHasOptionsMenu(true)
        (activity as AppCompatActivity).supportActionBar?.let{ ab ->
            ab.setDisplayHomeAsUpEnabled(false)
            ab.setTitle(R.string.app_name)
        }
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        inflater.inflate(R.menu.peercast_menu, menu)
    }

    override fun onCreateContextMenu(menu: ContextMenu, v: View, menuInfo: ContextMenu.ContextMenuInfo?) {
        val info = menuInfo as? ExpandableListView.ExpandableListContextMenuInfo ?: return

        val type = ExpandableListView.getPackedPositionType(info.packedPosition)
        val gPos = ExpandableListView.getPackedPositionGroup(info.packedPosition)
        val cPos = ExpandableListView.getPackedPositionChild(info.packedPosition)
        val inflater = MenuInflater(v.context)

        if (type == ExpandableListView.PACKED_POSITION_TYPE_GROUP) {
            // チャンネル
            inflater.inflate(R.menu.channel_context_menu, menu)
            val ch = listAdapter.getGroup(gPos)
            menu.setHeaderTitle(ch.ch.info.name)
            //val mKeep = menu.findItem(R.id.menu_ch_keep)
            //mKeep.isChecked = ch.isStayConnected
        } else if (type == ExpandableListView.PACKED_POSITION_TYPE_CHILD) {
            // サーヴァント
            inflater.inflate(R.menu.servent_context_menu, menu)
            val svt = listAdapter.getChild(gPos, cPos)
            menu.setHeaderTitle(svt.remoteEndPoint.toString())
        }
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.menu_remove_all_channel -> {
                listAdapter.channels.filter { ch ->
                    ch.ch.status.localRelays + ch.ch.status.localDirects == 0
                }.forEach { ch ->
                    executeRpcCommand {
                        stopChannel(ch.ch.channelId)
                    }
                }

            }

            else -> return false
        }
        return true
    }

    private fun executeRpcCommand(f: suspend PeerCastRpcClient.()->Unit){
        lifecycleScope.launchWhenStarted {
            val client = viewModel.rpcClient.value
            if (client == null){
                Timber.w("service not connected")
                return@launchWhenStarted
            }
            kotlin.runCatching {
                client.f()
            }.onFailure {
                Timber.e(it)
            }
        }
    }

    override fun onContextItemSelected(item: MenuItem): Boolean {
        val info = item.menuInfo as ExpandableListView.ExpandableListContextMenuInfo
        val type = ExpandableListView.getPackedPositionType(info.packedPosition)
        val gPos = ExpandableListView.getPackedPositionGroup(info.packedPosition)
        val cPos = ExpandableListView.getPackedPositionChild(info.packedPosition)

        if (listAdapter.groupCount <= gPos)
            return false
        val ch = listAdapter.getGroup(gPos)

        return when (item.itemId) {

            R.id.menu_ch_disconnect -> {
                Timber.i("Disconnect channel: $ch")
                executeRpcCommand { stopChannel(ch.ch.channelId) }
                true
            }

            R.id.menu_ch_play -> {
                val intent = LibPeerCast.createStreamIntent(ch.ch.channelId, appPrefs.port)
                try {
                    startActivity(intent)
                } catch (e: ActivityNotFoundException) {
                    Snackbar.make(requireView(), e.toString(), Snackbar.LENGTH_LONG).also {
                        val color = ContextCompat.getColor(it.context, R.color.md_red_700)
                        it.setTextColor(color)
                    }
                }
                true
            }

            R.id.menu_ch_bump -> {
                Timber.i("Bump channel: $ch")
                executeRpcCommand {
                    bumpChannel(ch.ch.channelId)
                }
                true
            }

            R.id.menu_svt_disconnect -> {
                //直下切断
                val conn = listAdapter.getChild(gPos, cPos)
                Timber.i("Disconnect connection: $conn")
                executeRpcCommand {
                    stopChannelConnection(ch.ch.channelId, conn.connectionId)
                }
                true
            }

            else -> super.onContextItemSelected(item)
        }
    }
}
